kudzu.8
.TH kudzu 1 "Red Hat, Inc."
.UC 4
.SH NAME
kudzu \- detects and configures new and/or changed hardware on a system

.SH SYNOPSIS
.B kudzu
.SH DESCRIPTION
.B kudzu
detects and configures new and/or changed hardware on a system.
When started, kudzu detects the current hardware, and checks it
against a database stored in
.I /etc/sysconfig/hwconf,
if one exists. It then determines if any hardware has been
added or removed from the system. If so, it gives the users
the opportunity to configure any added hardware, and unconfigure
any removed hardware. It then updates the database in
.I /etc/sysconfig/hwconf.

If no previous database exists,
.B kudzu
attempts to determine what devices have already been configured,
by looking at
.I /etc/modprobe.conf,
.I /etc/sysconfig/network-scripts/, and 
.I /etc/X11/xorg.conf.

.SH OPTIONS
.TP
\fI--usage\fP
Show short usage message.
.TP
\fI--help, -?\fP
Print help information.
.TP 
\fI-q, --quiet\fP
Run 'quietly'; do only configuration that doesn't require user input.
.TP
\fI-s, --safe\fP
Do only 'safe' probes that won't disturb hardware. Currently, this
disables the serial probe, the DDC monitor probe, and the PS/2 probe.
.TP
\fI-t, --timeout [seconds]\fP
This sets the timeout for the initial dialog. If no key is pressed
before the timeout elapses,
.B kudzu
exits, and
.I /etc/sysconfig/hwconf
is not updated.
.TP
\fI-k, --kernel [version]\fP
When determining whether a module exists, use the specified kernel
version. (If this is not set, it defaults to the current kernel
version.) Do not specify suffixes such as 'smp' or 'summit'; these
are automatically searched.
.TP
\fI-b, --bus [bus]\fP
Only probe on the specified bus.
.TP
\fI-c, --class [class]\fP
Only probe for the specified class.
.TP
\fI-f, --file [file]\fP
Read hardware probe info from file
.I file
and do not do an actual probe.
.TP
\fI-p, --probe\fP
Print probe information to the screen, and do not actually configure
or unconfigure any devices.
.SH FILES
.TP
.I /etc/sysconfig/hwconf
Listing of current installed hardware.
.TP
.I /etc/sysconfig/kudzu
Configuration for the boot-time hardware probe. Set 'SAFE' to something
other than 'no' to force only safe probes.
.TP
.I /etc/modprobe.conf
Module configuration file.
.TP
.I /etc/sysconfig/network-scripts/ifcfg-*
Network interface configuration files.
.TP
.SH BUGS
The serial probe will disturb any currently in-use devices, and returns
odd results if used on machines acting as serial consoles. On some older
graphics cards, the DDC probe can do strange things.
.SH AUTHOR
Red Hat, Inc.


debian/changelog
kudzu-kanotix (1.1.119-1) unstable; urgency=low

  * New upstream release:
    - make sure changing network devices are properly caught (#141338)
  * bump standards version to 3.6.2.
  * adapt build-depends to debhelper 4.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 30 Jul 2005 16:11:12 +0200

kudzu-kanotix (1.1.118-1) unstable; urgency=low

  * new upstream version:
    - make sure ISAPNP devices don't end up as CLASS_UNSPEC (#163949)
    - fix usb - class/subclass/protocol are hex, not decimal (#160450)

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Mon, 25 Jul 2005 16:25:30 +0200

kudzu-kanotix (1.1.116.2-2) unstable; urgency=low

  * switch to dpatch instead of monolithic patching.
  * recompile with gcc-4.0.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Fri, 22 Jul 2005 01:01:29 +0200

kudzu-kanotix (1.1.116.2-1) unstable; urgency=low

  * new upstream version:
    - tweak synaptics detection for ALPS (#158062, <pnasrat@redhat.com>)

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 28 May 2005 13:16:42 +0200

kudzu-kanotix (1.1.116-1) unstable; urgency=low

  * new upstream version:
    - hack for nForce4 (#153176)
    - fix uninitialized memory use in pci probe
    - fix snd-powermac configuration (#151591)
    - fix some potential segfaults if the files read are empty (#145998)
    - don't write install lines for sound cards (not useful with udev),
      add a 'index' parameter for the acutal module, not just the alias
      (#133759)
    - support modules.pcimap class/class_mask, and generic entries too
      (#146213)

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat,  7 May 2005 11:53:04 +0200

kudzu-kanotix (1.1.112-1) unstable; urgency=low

  * new upstream version:
    - make up some fan devices for mac things that can only be found in
      the device-tree (#151661)
    - fix crash in vio.c if the viocd file is empty (#154905)
    - fix firewire crash (#153987, <danieldk@pobox.com>)
    - add PCI class for HD audio

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Mon, 18 Apr 2005 19:46:46 +0200

kudzu-kanotix (1.1.111-1) unstable; urgency=low

  * new upstream version:
    - handle -q/--quiet

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 19 Mar 2005 22:17:04 +0100

kudzu-kanotix (1.1.110-1) unstable; urgency=low

  * new upstream version:
    - fix reading of modules.pcimap when /usr/share/hwdata isn't there
      (#150123, <dlehman@redhat.com>)
    - support fddiX device naming (#109689)

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Fri,  4 Mar 2005 01:28:44 +0100

kudzu-kanotix (1.1.109-1) unstable; urgency=low

  * new upstream version:
    - use synaptics for ALPS touchpad (#149619, <pnasrat@redhat.com>).

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 26 Feb 2005 14:45:20 +0100

kudzu-kanotix (1.1.108-1) unstable; urgency=low

  * new upstream version. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 19 Feb 2005 18:59:07 +0100

kudzu-kanotix (1.1.107-1) unstable; urgency=low

  * new upstream version. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu,  3 Feb 2005 14:35:51 +0100

kudzu-kanotix (1.1.106-1) unstable; urgency=low

  * new upstream version.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat,  8 Jan 2005 00:51:04 +0100

kudzu-kanotix (1.1.104-1) unstable; urgency=low

  * new upstream release.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 18 Dec 2004 15:43:09 +0100

kudzu-kanotix (1.1.101-1) unstable; urgency=low

  * new upstream release. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu,  2 Dec 2004 10:59:11 +0100

kudzu-kanotix (1.1.100-1) unstable; urgency=low

  * new upstream release. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 25 Nov 2004 15:52:43 +0100

kudzu-kanotix (1.1.96-1) unstable; urgency=low

  * new upstream version. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 23 Nov 2004 23:55:40 +0100

kudzu-kanotix (1.1.95-2) unstable; urgency=low

  * minor fixes for the debian specific build process. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 14 Oct 2004 23:05:15 +0200

kudzu-kanotix (1.1.95-1) unstable; urgency=low

  * merge new upstream release. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 14 Oct 2004 22:19:30 +0200

kudzu-kanotix (1.1.93-2) unstable; urgency=low

  * merge new upstream version.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Mon, 11 Oct 2004 21:37:09 +0200

kudzu-kanotix (1.1.93-1) unstable; urgency=low

  * fix crash when there's no modprobe.conf/modules.conf.

 -- Bill Nottingham <notting@redhat.com>  Fri,  8 Oct 2004 00:00:00 +0000

kudzu-kanotix (1.1.92-2) unstable; urgency=low

  * merge new upstream version.
  * minor bugfixes. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sun,  3 Oct 2004 22:09:56 +0200

kudzu-kanotix (1.1.90-4) unstable; urgency=low

  * fork from kudzu-kanotix.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 28 Sep 2004 00:26:19 +0200

kudzu-knoppix (1.1.90-3) unstable; urgency=low

  * fix deadlock. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 23 Sep 2004 23:25:23 +0200

kudzu-knoppix (1.1.90-2) unstable; urgency=low

  * sync to new upstream version. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 23 Sep 2004 22:57:42 +0200

kudzu-knoppix (1.1.89-6) unstable; urgency=low

  * clean up, minor fixes. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Fri, 17 Sep 2004 23:27:29 +0000

kudzu-knoppix (1.1.89-5) unstable; urgency=low

  * fix upstream version. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Fri, 17 Sep 2004 22:18:51 +0000

kudzu-knoppix (1.1.89-4) unstable; urgency=low

  * fix kudzu-knoppix-vesa. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 14 Sep 2004 12:42:40 +0200

kudzu-knoppix (1.1.89-3) unstable; urgency=low

  * re- add kudzu-knoppix-vesa. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 14 Sep 2004 02:44:21 +0200

kudzu-knoppix (1.1.89-2) unstable; urgency=low

  * new upstream release, ported to Kanotix BH8.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 14 Sep 2004 00:07:24 +0200

kudzu-knoppix (1.1.72-2) unstable; urgency=low


kudzu.py
# Copyright 2000-2003 Red Hat, Inc
#
# This software may be freely redistributed under the terms of the GNU
# public license.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

import _kudzu
import sys

consts=('BUS_', 'CLASS_', 'PROBE_')

for n in _kudzu.__dict__.keys():
    for c in consts:
        if len(n) > len(consts) and n[:len(c)] == c:
            sys.modules[__name__].__dict__[n] = _kudzu.__dict__[n]

# The base device class

class device:
    def __init__(self,table):
        self.deviceclass=table["class"]
        self.bus=table["bus"]
        self.index=table["index"]
        self.detached=table["detached"]
        self.desc=table["desc"]
        self.driver=table["driver"]
        self.device=table["device"]
	if self.deviceclass == CLASS_NETWORK:
	    if table.has_key("hwaddr"):
		self.hwaddr = table["hwaddr"]
	    else:
	    	self.hwaddr = None

    # maintain backwards compatibility
    def __getitem__(self, which):
        if which == 0:
            return self.device
        if which == 1:
            return self.driver
        if which == 2:
            return self.desc
        raise IndexError
    
    def setIndex(self,index):
        self.index=index

    def setDetached(self,detached):
        self.detached=detached

    def setDesc(self,desc):
        self.desc=desc

    def setDriver(self,driver):
        self.driver=driver

    def __repr__(self):
        res=""
        res=res+"Desc:           %s\n" % (self.desc)
        res=res+"Driver:         %s\n" % (self.driver)
        res=res+"Device:         %s\n" % (self.device)
        return res

# One class for each structure in the kudzu include files

class scsiDevice(device):
    def __init__(self,table):
        device.__init__(self,table)
        self.host=table["host"]
        self.channel=table["channel"]
        self.id=table["id"]
        self.lun=table["lun"]

    def setHost(self,host):
        self.host=host

    def setChannel(self,channel):
        self.channel=channel

    def setId(self,id):
        self.id=id

    def setLun(self,lun):
        self.lun=lun

class serialDevice(device):
    def __init__(self,table):
        device.__init__(self,table)
        self.pnpdesc=table["pnpdesc"]
        self.pnpmodel=table["pnpmodel"]
        self.pnpcompat=table["pnpcompat"]
        self.pnpmfr=table["pnpmfr"]

    def setPnPMfr(self,mfr):
        self.pnpmfr=mfr

    def setPnpModel(self,model):
        self.pnpmodel=model

    def setPnPCompat(self,compat):
        self.pnpcompat=compat

    def setPnpDesc(self,desc):
        self.pnpdesc=desc
    

class ideDevice(device):
    def __init__(self,table):
        device.__init__(self,table)
        self.physical=table["physical"]
        self.logical=table["logical"]

    def setPhysical(self,cyls):
        self.physical=cyls

    def setLogical(self,cyls):
        self.logical=cyls


class keyboardDevice(device):
    def __init__(self,table):
        device.__init__(self,table)

class psauxDevice(device):
    def __init__(self,table):
        device.__init__(self,table)


class sbusDevice(device):
    def __init__(self,table):
        device.__init__(self,table)
        self.width=table["width"]
        self.height=table["height"]
        self.freq=table["freq"]
        self.monitor=table["monitor"]

    def setRes(self,height,width):
        self.width=width
        self.height=height

    def setFreq(self,freq):
        self.freq=freq

    def setMonitor(self,monitor):
        self.monitor=monitor

class parallellDevice(device):
    def __init__(self,table):
        device.__init__(self,table)
        self.pnpdesc=table["pnpdesc"]
        self.pnpmodel=table["pnpmodel"]
        self.pnpmodes=table["pnpmodes"]
        self.pnpmfr=table["pnpmfr"]
        if(table.has_key("xres")):
            self.printeres=(table["xres"],table["yres"])
            self.printercolor=table["color"]
            self.printerascii=table["ascii"]
            self.printeruniprint=table["uniprint"]
        else:
            self.printeres=""
            self.printercolor=""
            self.printerascii=""
            self.printeruniprint=""
            
    def setPnPModel(self,model):
        self.pnpmodel=model

    def setPnPMfr(self,mfr):
        self.pnpmfr=mfr

    def setPnPModes(self,modes):
        self.pnpmodes=modes

    def setPnPDesc(self,desc):
        self.pnpdesc=desc

    def setPrinterRes(self,x,y):
        self.printerres=(x,y)

    def setPrinterColor(self,color):
        self.printercolor=color

    def setPrinterASCII(self,ascii):
        self.printerascii=ascii

    def setPrinterUniPrint(self,uniprint):
        self.printeruniprint=uniprint

class pciDevice(device):
    def __init__(self,table):
        device.__init__(self,table)
        self.vendorId=table["vendorId"]
        self.deviceId=table["deviceId"]
        self.subVendorId=table["subVendorId"]
        self.subDeviceId=table["subDeviceId"]
        self.pciType=table["pciType"]
	self.pcibus=table["pcibus"]
	self.pcidev=table["pcidev"]
	self.pcifn=table["pcifn"]

    def setVendorId(self,id):
        self.vendorId=id

    def setDeviceId(self,id):
        self.deviceId=id

    def setPciType(self,type):
        self.pciType=type

    def setSubVendorId(self,id):
        self.subVendorId=id

    def setSubDeviceId(self,id):
        self.subDeviceId=id
            
class pcmciaDevice(device):
    def __init__(self,table):
        device.__init__(self,table)
        self.vendorId=table["vendorId"]
        self.deviceId=table["deviceId"]

    def setVendorId(self,id):
        self.vendorId=id

    def setDeviceId(self,id):
        self.deviceId=id

    def setFunction(self,id):
        self.function=id

    def setSlot(self,id):
        self.slot=id
            
class ddcDevice(device):
    def __init__(self,table):
        device.__init__(self,table)
        if table.has_key("id"):
	    self.id=table["id"]
	else:
	    self.id = None
        self.horizSyncMin=table["horizSyncMin"]
        self.horizSyncMax=table["horizSyncMax"]
        self.vertRefreshMin=table["vertRefreshMin"]
        self.vertRefreshMax=table["vertRefreshMax"]
        self.physicalWidth=table["physicalWidth"]
        self.physicalHeight=table["physicalHeight"]
        self.mem=table["mem"]

    def __repr__(self):
        res=""
        res=res+"Desc:           %s\n" % (self.desc)
        res=res+"Id:             %s\n" % (self.id)
        res=res+"Mem on card:    %d\n" % (self.mem)
        res=res+"HSync:          %d-%d\n" % (self.horizSyncMin,
                                             self.horizSyncMax)
        res=res+"VSync:          %d-%d\n" % (self.vertRefreshMin,
                                             self.vertRefreshMax)
        res=res+"Physical width: %d mm\n" % (self.physicalWidth)
        res=res+"Physical height: %d mm\n" % (self.physicalHeight)
        return res

    def setID(self,id):
        self.id=id

    def setHorizSyncRange(self,min,max):
        self.horizSyncMin=min
        self.horizSyncMax=max

    def setVertRefreshRange(self,min,max):
        self.vertRefreshMin=min
        self.vertRefreshMax=max

    def setMem(self,mem):
        self.mem=mem

    def setPhysicalWidth(self, width):
        self.physicalWidth = width
    
    def setPhysicalHeight(self, height):
        self.physicalHeight = height
    
    def setModes(self,intdata):
        self.modes=[]

        # This is of the format xyxyxy....
        for i in range(0,len(intdata),2):
            modes.append((intdata[i],intdata[i+1]))

class usbDevice(device):

    def __init__(self,deviceclass,bus):
        device.__init__(self,deviceclass,bus)
        self.usbclass=""
        self.usbsubclass=""
        self.usbprotocol=""

    def __init__(self,table):
        device.__init__(self,table)
        self.usbclass=table["usbclass"]
        self.usbsubclass=table["usbsubclass"]
        self.usbprotocol=table["usbprotocol"]
        self.usbbus=table["usbbus"]
        self.usblevel=table["usblevel"]
        self.usbport=table["usbport"]
        self.vendorId=table["vendorid"]
        self.deviceId=table["deviceid"]
        self.usbmfr=table["usbmfr"]        
        self.usbprod=table["usbprod"]        
        

    def setUsbClass(self,usbclass):
        self.usbclass=usbclass

    def setUsbSubClass(self,usbclass):
        self.usbsubclass=usbclass

    def setUsbProtocol(self,protocol):
        self.usbprotocol=protocol


# Call the probefunction and return an array of
# objects with device information

def probe(devclass,bus,mode):
    res=_kudzu.probe(devclass,bus,mode)
    res2=[]
    for i in range(0,len(res)):
        dev=res[i]
        bus=dev["bus"]
        if(bus==BUS_PCI):
            res2.append(pciDevice(dev))
        elif(bus==BUS_SBUS):
            res2.append(sbusDevice(dev))
        elif(bus==BUS_PSAUX):
            res2.append(psauxDevice(dev))
        elif(bus==BUS_SERIAL):
            res2.append(serialDevice(dev))
        elif(bus==BUS_PARALLEL):
            res2.append(parallellDevice(dev))
        elif(bus==BUS_SCSI):
            res2.append(scsiDevice(dev))
        elif(bus==BUS_IDE):
            res2.append(ideDevice(dev))
        elif(bus==BUS_KEYBOARD):
            res2.append(keyboardDevice(dev))
        elif(bus==BUS_DDC):
            res2.append(ddcDevice(dev))
        elif(bus==BUS_USB):
            res2.append(usbDevice(dev))
	elif(bus==BUS_PCMCIA):
	    res2.append(pcmciaDevice(dev))
        else:
            res2.append(device(dev))

    return res2
